R/dependencies.R

```diff
@@ -697,6 +718,94 @@ get_local_pkgs_from_config <- function() {
 }
 
 
+# Given a list of packages (and their directories in the cache) of the form
+# list(`repo @ host`=<<path>>) create a dependency graph using the R
+# description files.
+get_internal_dependencies_graph <- function(internal_dependencies,
+                                            direction = "upstream") {
+
+  # get the package names from the DESCRIPTION files
+  # (they may not be the same name as the repo name)
+  package_names <- vapply(internal_dependencies,
+    function(repo_dir) {
+      fdesc <- file.path(repo_dir, "DESCRIPTION")
+      dsc <- desc::desc(fdesc)
+      return(unname(dsc$get("Package")))
+    },
+    FUN.VALUE = character(1)
+  )
+
+  # add the hashed_repo_and_host as the names so there
+  # is a mapping between
+  names(package_names) <- names(internal_dependencies)
+
+  # get the imports, suggests, depends for each package
+  # from the package description files, filter for only
+  # the internal packages and name in form hashed_repo_and_host
+  deps <- lapply(internal_dependencies,
+    function(repo_dir) {
+      fdesc <- file.path(repo_dir, "DESCRIPTION")
+      dsc <- desc::desc(fdesc)
+      names(package_names[package_names %in% dsc$get_deps()$package])
+    }
+  )
+
+  names(deps) <- names(internal_dependencies)
+
+  res <- list()
+  if ("upstream" %in% direction) {
+    res[["upstream_deps"]] <- deps
+  }
+  if ("downstream" %in% direction) {
+    downstream_deps <- list()
+    for(x in names(deps)){
+      downstream_deps[[x]] <- character(0)
+      for(y in names(deps)){
+        if(x %in% deps[[y]]){
+          downstream_deps[[x]] <- c(downstream_deps[[x]], y)
+        }
+      }
+    }
+    res[["downstream_deps"]] <- downstream_deps
+  }
+
+  res
+
+}
+
+
 
+#' Update existing stage_dependencies yaml file
+#'
+#' Using the existing stage_dependencies yaml file
+#' 'graph' to define internal dependencies, update the
+#' project yaml file to include to include all direct
+#' (i.e. distance 1) upstream and downstream repos
+#' @inheritParams dependency_structure
+#' @export
+update_stageddeps_yaml_with_direct_deps <- function(project = ".", feature = NULL,
+                                                    local_repos = get_local_pkgs_from_config(),
+                                                    verbose = 0){
+  dep_table <- dependency_structure(
+    project = project,
+    feature = feature,
+    local_repos = local_repos,
+    return_table_only = TRUE,
+    verbose = verbose
+  )
 
+  current_repo <- list("repo" = dep_table[dep_table$type == "current",]$repo,
```

Comment:
I think we should refactor this so we can test it and add tests (or if we don't do it now we add it as an issue) - if written like this:

```
update_stageddeps_yaml_with_direct_deps <- function(){

  dep_table <- get_dependency_table() 
  yaml_string <- get_yaml_from_dependency_table(dep_table) #this returns a string 
  write yaml_string to file = file.path(project, STAGEDDEPS_FILENAME) #don't need write_yaml
}
```

Then we can unit test get_yaml_from_dependency_table by plugging in different (made up) dependency tables and check the yaml that comes out is correct

Reply:
done